server/src/main/java/com/linecorp/centraldogma/server/internal/api/MetadataApiService.java

```diff
@@ -0,0 +1,393 @@
+                                                  @RequestObject Author author) {
+        final User member = new User(urlDecode(memberId));
+        return mds.getMember(projectName, member)
+                  .thenCompose(unused -> mds.removeMember(author, projectName, member));
+    }
+
+    /**
+     * POST /metadata/{projectName}/tokens
+     *
+     * <p>Adds a {@link Token} to the specified {@code projectName}.
+     */
+    @Post("/metadata/{projectName}/tokens")
+    @Decorator(ProjectOwnersOnly.class)
+    public CompletionStage<Revision> addToken(@Param("projectName") String projectName,
+                                              @RequestObject IdentifierWithRole request,
+                                              @RequestObject Author author) {
+        final ProjectRole role = toProjectRole(request.role());
+        return mds.findTokenByAppId(request.id())
+                  .thenCompose(token -> mds.addToken(author, projectName, token.appId(), role));
+    }
+
+    /**
+     * PUT /metadata/{projectName}/tokens
+     *
+     * <p>Updates the {@link ProjectRole} of the {@link Token} of the specified {@code appId}
+     * in the specified {@code projectName}.
+     */
+    @Patch("/metadata/{projectName}/tokens/{appId}")
+    @ConsumeType("application/json-patch+json")
+    @Decorator(ProjectOwnersOnly.class)
+    public CompletionStage<Revision> updateTokenRole(@Param("projectName") String projectName,
+                                                     @Param("appId") String appId,
+                                                     @RequestObject JsonPatch jsonPatch,
+                                                     @RequestObject Author author) {
+        final ReplaceOperation operation = ensureSingleReplaceOperation(jsonPatch, "/role");
+        final ProjectRole role = ProjectRole.of(operation.value());
+        return mds.findTokenByAppId(appId)
+                  .thenCompose(token -> mds.updateTokenRole(author, projectName, token, role));
+    }
+
+    /**
+     * DELETE /metadata/{projectName}/tokens/{appId}
+     *
+     * <p>Removes the {@link Token} of the specified {@code appId} from the specified {@code projectName}.
+     */
+    @Delete("/metadata/{projectName}/tokens/{appId}")
+    @Decorator(ProjectOwnersOnly.class)
+    public CompletionStage<Revision> removeToken(@Param("projectName") String projectName,
+                                                 @Param("appId") String appId,
+                                                 @RequestObject Author author) {
+        return mds.findTokenByAppId(appId)
+                  .thenCompose(token -> mds.removeToken(author, projectName, token.appId()));
+    }
+
+    /**
+     * PUT /metadata/{projectName}/tokens
+     *
+     * <p>Updates the {@link PerRolePermissions} of the specified {@code projectName} and {@code repoName}.
+     */
+    @Put("/metadata/{projectName}/repos/{repoName}/perm/role")
```

Comment:
Put method without the identifier at the end looks weird to me...

How about `@Put("/metadata/{projectName}/rolePermission/{repoName}")`?
I think this is updating the `rolePermission` of the specific repository, so `repoName` should be at the last.
Also I think we do not have to define the path as same with the API updating `perUserPermissions` and `perTokenPermissions` since this has no identifier(Updating the whole permission).
WDYT?

Reply:
I agree with you. I also think POST is more reasonable. Thanks!